### PR TITLE
Allow pam stack to call ReleaseSession

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Comment by YHNdnzj** [comment]
Hmm, according to #8598, apparently the fix might be in the wrong direction? I.e. rather than allowing to call `ReleaseSession` after dropping the privilege, the privilege should be retained and we need an alternative approach to track and terminate the PAM session helper.

I think a bigger yet comprehensive rework should be the way to go (https://github.com/systemd/systemd/pull/31033#issuecomment-1904265197), but noone has yet reviewed the PR (including me, sorry, completely forgot about it).

**Comment by YHNdnzj** [comment]
cc @yuwata

File: src/login/logind-dbus.c [hunk 1]
```diff
@@ -1183,6 +1183,13 @@ static int method_release_session(sd_bus_message *message, void *userdata, sd_bu
         if (r < 0)
                 return r;
 
+        r = get_sender_session(m, message, /* consult_display= */ false, error, &sender_session);
+        if (r < 0)
+                return r;
+
+        if (session != sender_session)
```
Comment (dtardon) [inline src/login/logind-dbus.c:1193]: This means _only the owner_ can call `ReleaseSession` now, doesn't it?

Comment (keszybz) [reply src/login/logind-dbus.c:1193]: Yes. But I think that's appropriate. `ReleaseSession` and friends are documented as: "These calls should never be invoked directly by clients. Creating/closing sessions is exclusively the job of PAM and its pam_systemd(8) module."

**Comment by keszybz** [comment]
> Hmm, according to #8598, the fix might be in the wrong direction? I.e. rather than allowing to call `ReleaseSession` after dropping the privilege, the privilege should be retained and we need an alternative approach to track and terminate the PAM session helper.

Yes, this PR doesn't implement that part of the request from #8598. It was targeted at #28514 and fixes *that* issue. I like that it's a simple solution. It also partially fixes the problems described in the other ticket, but doesn't solve all of them, because that'd require running the code at elevated privilege level. I know that there are reasons to do this, but it seems a bit iffy too.

**Comment by keszybz** [comment]
If #31033 is merged, then we won't need this PR. So let's put it on hold for now.

**Comment by bluca** [comment]
> If #31033 is merged, then we won't need this PR. So let's put it on hold for now.

Actually that one looks a lot more complex, and it's still in draft too, so doesn't seem like a good candidate for v256. This one is much smaller, and seems like we could take it for v256, as it fixes an actual problem. Then the other one can be reworked on top of this for v257?

**Comment by keszybz** [comment]
I think this would be reasonable. I don't think we should merge https://github.com/systemd/systemd/pull/31033 at this point in the release. It is mostly likely a better solution, but it's also a very risky change. So yeah, I think we should merge this one here to solve the most visible issue.

**Review by bluca — APPROVED** [review]

**Comment by keszybz** [comment]
mkosi / ci (opensuse, tumbleweed) is busted because of git sha256.
noble-s390x seems to have suffered a testbed aneurysm. 
jammy-amd64 failed in TEST-04-JOURNAL.
noble-amd64 failed in TEST-31-DEVICE-ENUMERATION.
Looks all unrelated.

File: src/login/logind-dbus.c [hunk 2]
```diff
@@ -1186,6 +1186,13 @@ static int method_release_session(sd_bus_message *message, void *userdata, sd_bu
...
+                return sd_bus_error_set(error, BUS_ERROR_NOT_IN_CONTROL, "You are not in control of this session");
```
Comment (poettering) [inline src/login/logind-dbus.c:1194]: hmm, please define a new error for this. the NotInControl error is really about session controllers, but this here really is different. Please add a new error.

Or just return SD_BUS_ERROR_ACCESS_DENIED or so.

**Pull request merged**